File: go/core/circuit.go
package core

// DefaultFailureThreshold: open the circuit after 5 consecutive failed messages.
const DefaultFailureThreshold = 5

// DefaultCooldownMs is the default circuit-breaker cooldown before an open
// circuit transitions to half-open.
const DefaultCooldownMs int64 = 30_000

// InitialCircuit returns a fresh, closed circuit with the given cooldown.
func InitialCircuit(cooldownMs int64) CircuitRecord {
	return CircuitRecord{State: CircuitClosed, ConsecutiveFailures: 0, CooldownMs: cooldownMs}
}

// OnFailure applies a delivery failure. A failure in half-open re-opens
// immediately and resets the cooldown clock. Otherwise it increments the
// consecutive-failure count and opens once it reaches threshold.
func OnFailure(rec CircuitRecord, now int64, threshold int) CircuitRecord {
	if rec.State == CircuitHalfOpen {
		openedAt := now
		return CircuitRecord{
			State: CircuitOpen, ConsecutiveFailures: rec.ConsecutiveFailures + 1,
			OpenedAt: &openedAt, CooldownMs: rec.CooldownMs,
		}
	}
	consecutiveFailures := rec.ConsecutiveFailures + 1
	if consecutiveFailures >= threshold {
		openedAt := now
		return CircuitRecord{
			State: CircuitOpen, ConsecutiveFailures: consecutiveFailures,
			OpenedAt: &openedAt, CooldownMs: rec.CooldownMs,
		}
	}
	return CircuitRecord{State: CircuitClosed, ConsecutiveFailures: consecutiveFailures, OpenedAt: nil, CooldownMs: rec.CooldownMs}
}

// OnSuccess applies a delivery success: reset to closed.
func OnSuccess(rec CircuitRecord) CircuitRecord {
	return CircuitRecord{State: CircuitClosed, ConsecutiveFailures: 0, OpenedAt: nil, CooldownMs: rec.CooldownMs}
}

// ToHalfOpen transitions an open circuit whose cooldown elapsed into half-open.
func ToHalfOpen(rec CircuitRecord) CircuitRecord {
	rec.State = CircuitHalfOpen
	return rec
}

// CircuitGate is the result of CanAttempt.
type CircuitGate struct {
	Allow bool
	Probe bool
}

// CanAttempt reports whether a delivery may be attempted now:
//
//   - closed    -> allowed (not a probe)
//   - open      -> allowed only once the cooldown elapsed, as a single probe
//     (caller then calls ToHalfOpen)
//   - half-open -> allowed as the single probe
func CanAttempt(rec CircuitRecord, now int64) CircuitGate {
	switch rec.State {
	case CircuitClosed:
		return CircuitGate{Allow: true, Probe: false}
	case CircuitHalfOpen:
		return CircuitGate{Allow: true, Probe: true}
	default: // open
		openedAt := int64(0)
		if rec.OpenedAt != nil {
			openedAt = *rec.OpenedAt
		}
		elapsed := now - openedAt
		if elapsed >= rec.CooldownMs {
			return CircuitGate{Allow: true, Probe: true}
		}
		return CircuitGate{Allow: false, Probe: false}
	}
}

File: packages/core/src/policy/circuit.ts
/** @fileoverview Per-endpoint circuit-breaker state machine (§9.2, G13, D4). @module @anyhook/core */
import type { CircuitRecord } from '../types/endpoint.js';

/** D4: open the circuit after 5 consecutive failed messages. */
export const DEFAULT_FAILURE_THRESHOLD = 5;
export const DEFAULT_COOLDOWN_MS = 30_000;

/** A fresh, closed circuit. */
export function initialCircuit(cooldownMs = DEFAULT_COOLDOWN_MS): CircuitRecord {
  return { state: 'closed', consecutiveFailures: 0, cooldownMs };
}

/**
 * Apply a delivery failure. A failure in `half-open` re-opens immediately and resets the cooldown.
 * Otherwise increments the consecutive count and opens once it reaches `threshold`.
 */
export function onFailure(
  rec: CircuitRecord,
  now: number,
  threshold = DEFAULT_FAILURE_THRESHOLD,
): CircuitRecord {
  if (rec.state === 'half-open') {
    return { ...rec, state: 'open', consecutiveFailures: rec.consecutiveFailures + 1, openedAt: now };
  }
  const consecutiveFailures = rec.consecutiveFailures + 1;
  if (consecutiveFailures >= threshold) {
    return { ...rec, state: 'open', consecutiveFailures, openedAt: now };
  }
  return { ...rec, state: 'closed', consecutiveFailures, openedAt: undefined };
}

/** Apply a delivery success: reset to closed. */
export function onSuccess(rec: CircuitRecord): CircuitRecord {
  return { ...rec, state: 'closed', consecutiveFailures: 0, openedAt: undefined };
}

/** Transition an `open` circuit whose cooldown elapsed into `half-open`. */
export function toHalfOpen(rec: CircuitRecord): CircuitRecord {
  return { ...rec, state: 'half-open' };
}

/**
 * Whether a delivery may be attempted now.
 * - `closed`    → allowed (not a probe)
 * - `open`      → allowed only once the cooldown elapsed, as a single probe (caller then `toHalfOpen`)
 * - `half-open` → allowed as the single probe
 */
export function canAttempt(rec: CircuitRecord, now: number): { allow: boolean; probe: boolean } {
  if (rec.state === 'closed') return { allow: true, probe: false };
  if (rec.state === 'half-open') return { allow: true, probe: true };
  // open
  const elapsed = now - (rec.openedAt ?? 0);
  if (elapsed >= rec.cooldownMs) return { allow: true, probe: true };
  return { allow: false, probe: false };
}
